---
title: "Top N"
author: "Caleb Easterly"
date: "April 26, 2018"
output:
  html_document: default
  word_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The goal here is to estimate the top *n* 'things' that are changing across WS and NS, by fold change. The fold change is calculated by the sum of the spectral counts attributed to each 'thing' - in eggNOG mapper and BLAST, the counts are aggregated by protein; in MEGAN, they are aggregated by eggNOG orthologous group; and in metaGOmics and Unipept, they are assigned to GO terms. In all cases, the Laplace correction is made, which is just adding 1 to all observed counts. This prevents division by zero, and allows for fold change estimation when one 'thing' was seen in one sample but not in another. 

```{r}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("GO.db")
```

# Reading in data
Necessary packages:
```{r message=FALSE}
library(GO.db)
library(dplyr)
library(kableExtra)
library(data.table)
```

## Peptides

These are the counts used in eggNOG-mapper and BLAST results interpretation. 

```{r}
peptidesUPS1 <- read.delim("peptide_shaker_outputs/UPS/UPS1_03_[Peptide_Shaker_on_data_21__Peptide_Report].tabular",
                         stringsAsFactors = FALSE) %>%
    select(peptide = Sequence, countUPS1 = "X.Validated.PSMs")

peptidesUPS2 <- read.delim("peptide_shaker_outputs/UPS/UPS2_03_[Peptide_Shaker_on_data_25__Peptide_Report].tabular",
                         stringsAsFactors = FALSE) %>%
    select(peptide = Sequence, countUPS2 = "X.Validated.PSMs") 

peptides_all <- full_join(peptidesUPS1, peptidesUPS2, by = "peptide")
peptides_all[is.na(peptides_all)] <- 0
```

## eggNOG mapper results

Join peptides to counts, and calculate ratios.

```{r}
eggnog <- read.delim("eggnogmap_results/UPS/Galaxy14-[eggNOG_Mapper_on_data_12__diamond.annotations].tabular",
                     stringsAsFactors = FALSE,
                     header = FALSE) %>% 
    select(peptide = V1, protein = V2, gene = V5, go = V6, ko = V7, desc = V13)
eggnog
eggnog_w_counts <- left_join(eggnog,
                             peptides_all,
                             by = "peptide") %>%
    group_by(protein, desc, gene) %>%
    summarize(sumCountUPS2 = sum(countUPS2) + 1, sumCountUPS1 = sum(countUPS1) + 1) %>%
    mutate(log2ratio = log2(sumCountUPS2/sumCountUPS1)) %>%
    arrange(-log2ratio) %>%
    select(protein, gene, sumCountUPS2, sumCountUPS1, log2ratio, desc)
```


Top *N*:
```{r echo=FALSE, results='asis'}
library(knitr)
n <- 100
kable(head(eggnog_w_counts, n), caption = paste("eggNOG-mapper: top", n, "fold changes"))
```

The <a href="http://pfam.xfam.org/family/PF05658">Pfam:YadA</a> refers to the YadA head domain in the trimeric autotransporter adhesin protein family. 

## UPS
```{r}
ups <- read.delim('uniprot-UPS-go-terms.tab',
                  stringsAsFactors = FALSE) %>%
    select(GO.terms = Gene.ontology.IDs)
ups$GO.terms <- sapply(ups$GO.terms, function( x ) { paste(Term(unlist(strsplit(x, '; '))), collapse = "; ") })
ups
```


## Blast
```{r}
blast <- read.delim('blast_results/blastp_vs_nr_current.tabular',
                    stringsAsFactors = FALSE,
                    header = FALSE) %>%
    select(peptide = V1, protein = V2, desc = V25)

blast_counts <- left_join(blast, peptides_all, by = "peptide") %>%
    group_by(protein, desc) %>%
    filter(!is.na(countWS) & !is.na(countNS)) %>%
    summarize(sumCountWS = sum(countWS) + 1, sumCountNS = sum(countNS) + 1) %>%
    mutate(log2ratio = log2(sumCountWS/sumCountNS)) %>%
    arrange(-log2ratio)
```

Print results:
```{r echo=FALSE, results='asis'}
kable(head(blast_counts, n), caption = paste("BLAST: Top ", n, " fold changes"))
```

## metaGOmics

We take the top *n* results that have a FDR-corrected $q$ value less than 0.05. 

```{r}
metagomics <- read.delim("metaGOmics_results/UPS/MetaGOmics_Compare_UPS2_UPS1.txt",
                         comment.char = "#") %>%
    select(go = GO.acc,
           name = GO.name,
           log2ratio =  Laplace.corr..Log.2..fold.change,
           p = Laplace.corr..q.value)

metagomics_filt <- metagomics %>% filter(p < 0.05) %>%
    arrange(-log2ratio)
```

Top *n* results:
```{r echo=FALSE, results='asis'}
kable(head(metagomics_filt, n), caption = paste("MetaGOmics: Top", n, "fold changes"))
```

## MEGAN

```{r}
megan <- read.delim("MEGAN_outputs/737NSvsWS_EGGNOGcount.csv") %>%
    select(og = X.Datasets, countNS = X737_NS_BLASTOutput_2StepCombined,
           countWS = X737_WS_BLASTOutput_2StepCombined) %>%
    mutate(corrWS = countWS + 1,
           corrNS = countNS + 1,
           log2ratio = log2(corrWS/corrNS)) %>%
    arrange(-log2ratio)
```

Top *n* results:
```{r echo=FALSE, results='asis'}
kable(head(megan, n), caption = paste("MEGAN: Top", n, "fold changes"))
```

Note that <a href="https://enzyme.expasy.org/EC/3.2.1.52">EC 3.2.1.52</a> is a beta-hexosaminidase and <a href="http://eggnogdb.embl.de/#/app/results?target_nogs=ENOG410YESU">ENOG410YESU</a> is involved in cell wall/membrane/envelope biogenesis.

## Unipept

```{r}
unipept_results_NS <- paste('unipept_results/',
                               list.files("unipept_results/", pattern = "^737NS.*\\.csv"),
                               sep = "")
unipept_results_WS <- paste('unipept_results/',
                               list.files("unipept_results/", pattern = "^737WS.*\\.csv"),
                               sep = "")

unipeptNS <- lapply(unipept_results_NS, function(i) {
        read.delim(i, sep = ',', as.is = TRUE)}) %>%
    bind_rows() %>%
    select(-X) %>%
    rename(peptides = X.peptides)

unipeptWS <- lapply(unipept_results_WS, function(i) {
    read.delim(i, sep = ',', as.is = TRUE)}) %>%
    bind_rows() %>%
    select(-X) %>%
    rename(peptides = X.peptides)

unipept_all <- inner_join(unipeptNS, unipeptWS, by = c("GO.term", "Name")) %>%
    mutate(lapCountNS = peptides.x + 1, lapCountWS = peptides.y + 1,
           log2ratio = log(lapCountWS/lapCountNS)) %>%
    select(GO.term, Name, lapCountWS, lapCountNS, log2ratio) %>%
    arrange(-log2ratio)
```
```{r}
unipeptUPS1 <- read.delim("unipept_results/UPS/UPS1_Galaxy19-[Unipept_pept2go_on_data_16_GO_tsv].tabular", header=FALSE) %>%
    select(GO.term = V3) %>%
    group_by(GO.term) %>%
    count()
unipeptUPS2 <- read.delim("unipept_results/UPS/UPS2_Galaxy23-[Unipept_pept2go_on_data_17_GO_tsv].tabular", header=FALSE) %>%
    select(GO.term = V3) %>%
    group_by(GO.term) %>%
    count()
unipept_all <- inner_join(unipeptUPS1, unipeptUPS2, by = "GO.term") %>%
    mutate(lapCountUPS1 = n.x + 1, lapCountUPS2 = n.y + 1,
           log2ratio = log(lapCountUPS2/lapCountUPS1)) %>%
    select(GO.term, lapCountUPS2, lapCountUPS1, log2ratio) %>%
    arrange(-log2ratio)
unipept_all$Name <- Term(unipept_all$GO.term)
unipept_all
```

Top *n*:
```{r echo=FALSE, results='asis'}
kable(head(unipept_all, n), caption = paste("Unipept: top", n, "Fold Changes"))
```


## MetaProteomeAnalyzer

```{r}
ns_files <- list.files('mpa_results/NS', pattern = "_proteins.csv", full.names = TRUE)
pNS <- bind_rows(lapply(ns_files, function(i) read.delim(i, stringsAsFactors = FALSE))) %>%
    group_by(Protein.Accession, Protein.Description) %>%
    summarize("SpecCount" = sum(Spectral.Count))
ws_files <- list.files('mpa_results/WS', pattern = "_proteins.csv", full.names = TRUE)
pWS <- bind_rows(lapply(ws_files, function(i) read.delim(i, stringsAsFactors = FALSE))) %>%
    group_by(Protein.Accession, Protein.Description) %>%
    summarize("SpecCount" = sum(Spectral.Count))

pAll <- full_join(pNS, pWS, by=c("Protein.Accession", "Protein.Description")) %>%
    mutate(nsCount = SpecCount.x + 1, wsCount = SpecCount.y + 1,
           log2ratio = log2(wsCount/nsCount)) %>%
    arrange(-log2ratio)

```

Top *n*:
```{r echo=FALSE, results='asis'}
kable(head(pAll, n), caption = paste("MetaProteomeAnalyzer: top", n, "Fold Changes"))
```


## Prophane

```{r}
prophane_WS <- read.delim("prophane_results/summary_WS.txt", sep = "\t",
                          stringsAsFactors = FALSE) %>% select(members_identifier,
                                                               ws = quant..sample_737..mean.,
                                                               ws_go1 = task_6..fun_from_TIGRFAMs_15_cut_tc..tigrfam2GO,
                                                               ws_go2 = task_7..fun_from_PFAMs_32..pfam2GO,
                                                               ws_go3 = task_8..fun_from_eggNog_4.5.1..og2GO)
prophane_NS <- read.delim("prophane_results/summary_NS.txt", sep = "\t",
                          stringsAsFactors = FALSE) %>% select(members_identifier,
                                                               ns = quant..sample_737..mean.,
                                                               ns_go1 = task_6..fun_from_TIGRFAMs_15_cut_tc..tigrfam2GO,
                                                               ns_go2 = task_7..fun_from_PFAMs_32..pfam2GO,
                                                               ns_go3 = task_8..fun_from_eggNog_4.5.1..og2GO)
prophane <- data.table(inner_join(prophane_NS, prophane_WS, by = c("members_identifier")) %>%
    filter(ns != "-", ws != "-") %>%
    mutate(fc = log2(as.double(ws)/as.double(ns)),
           gos = gsub(" ","", paste(ns_go1, ws_go1, ns_go2, ws_go2, ns_go3, ws_go3, sep=","))) %>%
    filter(gos != "") %>%
    arrange(-fc) %>%
    select(fc, gos, ns_go1, ws_go1, ns_go2, ws_go2, ns_go3, ws_go3))
prophane$gos <- sapply(strsplit(prophane$gos,","), function(x) {paste(unique(x[x != "-"]), collapse = ",")})
prophane <- prophane[ , list( gos = unlist( strsplit( gos , "," ) ) ) , by = fc ]
prophane$names <- Term(prophane$gos)
prophane <- prophane %>% group_by_at(vars(fc)) %>%
    summarize_all(paste, collapse="; ")
```


## Unipept and MetaGOmics

This is comparing the estimated fold changes for GO terms for the two tools that natively use GO terms, Unipept and MetaGOmics. 

```{r}
um <- inner_join(unipept_all, metagomics, by = c("GO.term" = "go"))
png("um_cor.png", res = 700, width = 4, height = 4, units = "in")
plot(log2ratio.x ~ log2ratio.y, data = um, pch = 20, cex = 0.3,
     xlab = "Unipept: Log2 Ratio",
     ylab = "MetaGOmics: Log2 Ratio")
mod <- lm(log2ratio.x ~ log2ratio.y, data = um)
abline(0, 1, col = "red", lty = 2)
abline(coef(mod))
grid()
dev.off()
cor.test(um$log2ratio.x, um$log2ratio.y,
         method = "pearson", alternative = "two.sided")
```

### Rank comparison

How do metaGOmics and Unipept rankings compare? 

```{r}
top5_meta <- metagomics_filt$go[1:5]
m2u_ranks <- rep(0, 5)
u2m_ranks <- rep(0, 5)
top5_uni <- unipept_all$GO.term[1:5]
for (i in 1:5){
    m2u_rank <- which(unipept_all$GO.term == top5_meta[i])
    u2m_rank <- which(metagomics_filt$go == top5_uni[i])
    m2u_ranks[i] <- ifelse(length(m2u_rank) == 1, m2u_rank, NA)
    u2m_ranks[i] <- ifelse(length(u2m_rank) == 1, u2m_rank, NA)
}

m2u <- data.frame("go_term" = top5_meta,
                  "rank_in_metagomics" = 1:5,
                  "rank_in_unipept" = m2u_ranks)

u2m <- data.frame("go_term" = top5_uni,
                  "rank_in_unipept" = 1:5,
                  "rank_in_metagomics" = u2m_ranks)

```

### How Unipept Ranks MetaGOmics' top 5
```{r echo=FALSE, results='asis'}
kable(m2u)
```

### How MetaGOmics ranks Unipept's top 5
```{r echo=FALSE, results='asis'}
kable(u2m)
```

## Top n Table
```{r}
rank_table <- data.frame(eggNOG = head(eggnog_w_counts$desc, n),
                          MEGAN = head(megan$og, n),
                     metaGOmics = head(metagomics_filt$name, n),
                            MPA = head(pAll$Protein.Description, n),
                        Unipept = head(unipept_all$Name, n))

# Remove extra text
rank_table$MPA <- sapply(rank_table$MPA, function( x ) { gsub('[A-Z0-9_]+ (.+) OS.+','\\1', x) })
rank_table$MEGAN <-sapply(rank_table$MEGAN, function( x ) { gsub('[A-Z0-9]+ (.+)','\\1', x) })

write.table(rank_table, "top100-UPS.tabular", sep="\t", quote = TRUE)
```


```{r}
n <- 100
functool1 <- as.character(rank_table[1:n,1])
functool2 <- as.character(rank_table[1:n,2])
functool3 <- as.character(rank_table[1:n,3])
functool4 <- as.character(rank_table[1:n,4])
functool5 <- as.character(rank_table[1:n,5])

removeIntraduplicates <- function(col){
    while(length(col[duplicated(col)]) > 0){
        col[duplicated(col)] <- as.character(paste(col[duplicated(col)], " "))
    }
    col
}

removeInterduplicates <- function(col1, col2){
    as.character(sapply(col2, function(x) ifelse(x %in% col1, paste(x, "  "), x)))
}

functool1 <- removeIntraduplicates(functool1)
functool2 <- removeIntraduplicates(functool2)
functool3 <- removeIntraduplicates(functool3)
functool4 <- removeIntraduplicates(functool4)
functool5 <- removeIntraduplicates(functool5)

functool2 <- removeInterduplicates(functool1, functool2)
functool3 <- removeInterduplicates(functool1, functool3)
functool3 <- removeInterduplicates(functool2, functool3)
functool4 <- removeInterduplicates(functool1, functool4)
functool4 <- removeInterduplicates(functool2, functool4)
functool4 <- removeInterduplicates(functool3, functool4)
functool5 <- removeInterduplicates(functool1, functool5)
functool5 <- removeInterduplicates(functool2, functool5)
functool5 <- removeInterduplicates(functool3, functool5)
functool5 <- removeInterduplicates(functool4, functool5)

getMatches <- function(col1, col2){
    matches_col1 <- as.character(sapply(col2, function (x){
        trimmed_GO <- gsub("\\s+$","", x)
        trimmed_col1 <- lapply(col1, function (x) gsub("\\s+$","", x))
        ifelse(trimmed_GO %in% trimmed_col1,
               col1[trimmed_GO == trimmed_col1],
               FALSE)
    }))
    links <- data.frame("source" = matches_col1,
                        "target" = col2)
    links[links$source != FALSE,]
}

getPseudomatches <- function(col1, col2){
    matches_col1 <- as.character(sapply(col2, function (x){
        trimmed_GO <- gsub("\\s+$","", x)
        trimmed_col1 <- lapply(col1, function (x) gsub("\\s+$","", x))
        forward_pseudomatch <- grepl(tolower(trimmed_GO), tolower(trimmed_col1), fixed = TRUE)
        reverse_pseudomatch <- sapply(trimmed_col1, function (x) grepl(x, trimmed_GO))
        ifelse((TRUE %in% forward_pseudomatch | TRUE %in% reverse_pseudomatch)
               & !trimmed_GO %in% trimmed_col1,
               ifelse(TRUE %in% forward_pseudomatch,
                      col1[forward_pseudomatch],
                      ifelse(col1[reverse_pseudomatch] == "", FALSE, col1[reverse_pseudomatch])),
               FALSE)
    }))
    links <- data.frame("source" = matches_col1,
                        "target" = col2)
    links[links$source != FALSE,]
}


matches <- rbind(
    getMatches(functool1, functool2),
    getMatches(functool1, functool3),
    getMatches(functool2, functool3),
    getMatches(functool1, functool4),
    getMatches(functool2, functool4),
    getMatches(functool3, functool4),
    getMatches(functool1, functool4),
    getMatches(functool2, functool4),
    getMatches(functool3, functool4),
    getMatches(functool1, functool5),
    getMatches(functool2, functool5),
    getMatches(functool3, functool5),
    getMatches(functool4, functool5)
)

pseudoMatches <- rbind(
    getPseudomatches(functool1, functool2),
    getPseudomatches(functool1, functool3),
    getPseudomatches(functool2, functool3),
    getPseudomatches(functool1, functool4),
    getPseudomatches(functool2, functool4),
    getPseudomatches(functool3, functool4),
    getPseudomatches(functool1, functool4),
    getPseudomatches(functool2, functool4),
    getPseudomatches(functool3, functool4),
    getPseudomatches(functool1, functool5),
    getPseudomatches(functool2, functool5),
    getPseudomatches(functool3, functool5),
    getPseudomatches(functool4, functool5)
)
```



```{r}
library(networkD3)
library(tidyverse)

values <- c()
values[1:(length(functool1) * 4)] <- 1
values[(length(functool1) * 4 + 1):(length(functool1) * 4 + nrow(matches) + nrow(pseudoMatches))] <- .1
#values <- "1"

# Make a connection data frame
links=data.frame(
    source=c(as.character(functool1), as.character(functool2), as.character(functool3), as.character(functool4), as.character(matches$source), as.character(pseudoMatches$source)),
    target=c(as.character(functool2), as.character(functool3), as.character(functool4), as.character(functool5), as.character(matches$target), as.character(pseudoMatches$target)),
    group=as.factor(c(rep("invisible", length(functool1) * 4), rep("matches", nrow(matches)), rep("pseudomatches", nrow(pseudoMatches)))),
    value= c(values))
# A3, D2, .5
links
functool1
 
# From these flows we need to create a node data frame: it lists every entities involved in the flow
nodes=data.frame(name=c(as.character(links$source), as.character(links$target)) %>% unique())
links$IDsource=match(links$source, nodes$name)-1 
links$IDtarget=match(links$target, nodes$name)-1

nodes$group = as.factor(c(rep("tool1", length(functool1)),
                          rep("tool2", length(functool2)),
                          rep("tool3", length(functool3)),
                          rep("tool4", length(functool4)),
                          rep("tool5", length(functool5))))

my_color <- 'd3.scaleOrdinal() .domain(["invisible", "matches", "pseudomatches", "tool1", "tool2", "tool3", "tool4", "tool5"]) .range(["white", "blue", "orange", "green", "purple", "red", "yellow", "black"])'
 
# Make the Network. I call my colour scale with the colourScale argument
n <- sankeyNetwork(Links = links, Nodes = nodes, Source = "IDsource", Target = "IDtarget",
                   Value = "value", NodeID = "name", colourScale=my_color, LinkGroup="group", NodeGroup="group",
                   iterations = 0)
n
saveNetwork(n, "top100sankeyR-UPS.html", selfcontained = TRUE)
```
```{r}
rep("D", 5)
```


